src/replay/execution.zig

```diff
@@ -0,0 +1,198 @@
+const sig = @import("../sig.zig");
+const replay = @import("lib.zig");
+
+const core = sig.core;
+
+const Allocator = std.mem.Allocator;
+
+const ThreadPool = sig.sync.ThreadPool;
+
+const Pubkey = core.Pubkey;
+const Slot = core.Slot;
+
+const AccountsDB = sig.accounts_db.AccountsDB;
+const BlockstoreReader = sig.ledger.BlockstoreReader;
+
+const ForkProgress = sig.consensus.progress_map.ForkProgress;
+const ProgressMap = sig.consensus.ProgressMap;
+
+const ConfirmSlotFuture = replay.confirm_slot.ConfirmSlotFuture;
+const EpochTracker = replay.trackers.EpochTracker;
+const SlotTracker = replay.trackers.SlotTracker;
+
+const confirmSlot = replay.confirm_slot.confirmSlot;
+
+/// State used for replaying and validating data from blockstore/accountsdb/svm
+pub const ReplayExecutionState = struct {
+    allocator: Allocator,
+    logger: sig.trace.ScopedLogger("replay-execution"),
+    my_identity: Pubkey,
+    vote_account: ?Pubkey,
+    accounts_db: *AccountsDB,
+    thread_pool: *ThreadPool,
+    blockstore_reader: *BlockstoreReader,
+    slot_tracker: *SlotTracker,
+    epochs: *EpochTracker,
+    progress_map: ProgressMap,
+
+    pub fn init(
+        allocator: Allocator,
+        logger: sig.trace.Logger,
+        my_identity: Pubkey,
+        thread_pool: *ThreadPool,
+        accounts_db: *AccountsDB,
+        blockstore_reader: *BlockstoreReader,
+        slot_tracker: *SlotTracker,
+        epochs: *EpochTracker,
+    ) Allocator.Error!ReplayExecutionState {
+        return .{
+            .allocator = allocator,
+            .logger = .from(logger),
+            .my_identity = my_identity,
+            .vote_account = null, // voting not currently supported
+            .accounts_db = accounts_db,
+            .thread_pool = thread_pool,
+            .blockstore_reader = blockstore_reader,
+            .slot_tracker = slot_tracker,
+            .epochs = epochs,
+            .progress_map = ProgressMap.INIT,
+        };
+    }
+
+    pub fn deinit(self: ReplayExecutionState) void {
+        self.epochs.deinit(self.allocator);
+    }
+};
+
+/// 1. Replays transactions from all the slots that need to be replayed.
+/// 2. Store the replay results into the relevant data structures.
+///
+/// Analogous to [replay_active_banks](https://github.com/anza-xyz/agave/blob/3f68568060fd06f2d561ad79e8d8eb5c5136815a/core/src/replay_stage.rs#L3356)
+pub fn replayActiveSlots(state: *ReplayExecutionState) !bool {
+    const active_slots = try state.slot_tracker.activeSlots(state.allocator);
+    if (active_slots.len == 0) {
+        return false;
+    }
+
+    var slot_statuses = std.ArrayListUnmanaged(struct { Slot, ReplaySlotStatus }){};
+    defer {
+        for (slot_statuses.items) |status| status[1].confirm.destroy(state.allocator);
+        slot_statuses.deinit(state.allocator);
+    }
+    for (active_slots) |slot| {
+        const result = try replaySlot(state, slot);
+        try slot_statuses.append(state.allocator, .{ slot, result });
```

Comment:
This seems like a leak, shouldn't there be a `errdefer result.destroy();` here?

Reply:
Maybe I changed this after your comment. The current code looks like this:
```zig
    var slot_statuses = std.ArrayListUnmanaged(struct { Slot, ReplaySlotStatus }).empty;
    defer {
        for (slot_statuses.items) |status| status[1].confirm.destroy(state.allocator);
        slot_statuses.deinit(state.allocator);
    }
    for (active_slots) |slot| {
        const result = try replaySlot(state, slot);
        try slot_statuses.append(state.allocator, .{ slot, result });
    }
```